### PR TITLE
Promote ReplicaSet Pod quota limit

### PR DESCRIPTION
**Pull request by BobyMCbobs**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Promotes the following E2E test to Conformance:
test/e2e/apps/replica_set.go: "should surface a failure condition on a common issue like exceeded quota"

**Special notes for your reviewer**:
Part of Umbrella Issue #78747.  
[Testgrid results](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=should%20surface%20a%20failure%20condition%20on%20a%20common%20issue%20like%20exceeded%20quota&include-filter-by-regex=ReplicaSet%20should%20surface%20a%20failure%20condition%20on%20a%20common%20issue%20like%20exceeded%20quota&width=5).  
Test(s):
- [sig-apps] ReplicaSet should surface a failure condition on a common issue like exceeded quota

Endpoint(s):
- replaceAppsV1NamespacedReplicaSet

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
/area conformance
/area test
@kubernetes/sig-apps-pr-reviews
@kubernetes/sig-architecture-pr-reviews
@kubernetes/cncf-conformance-wg

### CODE REVIEW

File: test/e2e/apps/replica_set.go [hunk 1]
```diff
@@ -102,7 +102,14 @@ var _ = SIGDescribe("ReplicaSet", func() {
 		testReplicaSetServeImageOrFail(f, "private", privateimage.GetE2EImage())
 	})
 
-	ginkgo.It("should surface a failure condition on a common issue like exceeded quota", func() {
+	/*
+			    Release : v1.16
```
Comment (draveness) [inline test/e2e/apps/replica_set.go]: Please reformat this comments.

Comment (BobyMCbobs) [reply test/e2e/apps/replica_set.go]: The formatting has been updated. Hopefully it's better :smiley:

File: test/e2e/apps/replica_set.go [hunk 2]
```diff
@@ -102,7 +102,14 @@ var _ = SIGDescribe("ReplicaSet", func() {
 		testReplicaSetServeImageOrFail(f, "private", privateimage.GetE2EImage())
 	})
 
-	ginkgo.It("should surface a failure condition on a common issue like exceeded quota", func() {
+	/*
+		Release : v1.16
+		Testname: Validate ReplicaSet pod limit restrictions
+		Description: Apply quota to the namespace, prohibiting more than two pods.
```
Comment (hh) [inline test/e2e/apps/replica_set.go]: Description needs some work, note the difference in:
`create pod with three replicas MUST fail'
vs
`when creating a replicaset of 3, replicasetConditions MUST include the failure X`

**Comment by mgdevstack** [comment]
+1 for https://github.com/kubernetes/kubernetes/pull/74891#issuecomment-479234581 and It's good to rewrite e2e.
>So the test should also verify that the condition is true when the replica can't be created and verify the condition is either removed OR the status is set to false when the condition does not occur.
---
I'm not sure, If it has been picked up to update documentation based on https://github.com/kubernetes/kubernetes/pull/74891#issuecomment-486272557

**Comment by timothysc** [comment]
/assign @timothysc 

Please address current PR comments.

File: test/e2e/apps/replica_set.go (hunk 2 above)
Comment (BobyMCbobs) [reply test/e2e/apps/replica_set.go]: The description has been updated to be more specific :smile:

**Comment by devaips** [comment]
@timothysc

It's not clear to me what documentation [you were referring to](https://github.com/kubernetes/kubernetes/pull/74891#issuecomment-486272557) in the previous PR for this test. The [Conformance Test Requirements](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-requirements) doc?

Looking at that past PR I also do not fully understand why it was not merged. I see where @bgrant0607 voiced their [concerns](https://github.com/kubernetes/kubernetes/pull/74891#issuecomment-479234581), but it was also approved after that. Is it to be understood that this test needs to be rewritten?

**Review by timothysc — APPROVED** [review]
/lgtm 
/assign @bgrant0607 @smarterclayton 
^ for approval. 

Even though it depends on conditions it specifically is checking ReplicaSetCondition.ReplicaSetConditionType which only checks for hard failure from what read.

File: test/conformance/testdata/conformance.txt [hunk 3]
```diff
@@ -36,6 +36,7 @@ test/e2e/apps/statefulset.go: "should perform canary updates and phased rolling
 test/e2e/apps/statefulset.go: "Scaling should happen in predictable order and halt if any stateful pod is unhealthy"
 test/e2e/apps/statefulset.go: "Burst scaling should run to completion even with unhealthy pods"
 test/e2e/apps/statefulset.go: "Should recreate evicted statefulset"
+test/e2e/apps/replica_set.go: "should surface a failure condition on a common issue like exceeded quota"
```
Comment (spiffxp) [inline test/conformance/testdata/conformance.txt]: this is listed on the wrong line

**Review by spiffxp — CHANGES_REQUESTED** [review]
the pull-kubernetes-bazel-test failure is legitimate, please address

I don't think this should block `/approve` but it's something that needs to be addressed in order to merge

File: test/conformance/testdata/conformance.txt (hunk 3 above)
Comment (spiffxp) [reply test/conformance/testdata/conformance.txt]: see https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/80148/pull-kubernetes-bazel-test/1154125960350535681#0:build-log.txt%3A56

**Comment by spiffxp** [comment]
needs rebase

**Comment by spiffxp** [comment]
Test name is in the wrong location in conformance.txt
```
INFO: From Testing //test/conformance:conformance_test:
==================== Test output for //test/conformance:conformance_test:
--- test/conformance/testdata/conformance.txt	2019-07-29 20:31:40.322864380 +0000
+++ test/conformance/conformance.txt	2019-07-29 20:31:40.322864380 +0000
@@ -33,8 +33,8 @@
 test/e2e/apps/rc.go: "should adopt matching pods on creation"
 test/e2e/apps/rc.go: "should release no longer matching pods"
 test/e2e/apps/replica_set.go: "should serve a basic image on each replica with a public image"
-test/e2e/apps/replica_set.go: "should adopt matching pods on creation and release no longer matching pods"
 test/e2e/apps/replica_set.go: "should surface a failure condition on a common issue like exceeded quota"
+test/e2e/apps/replica_set.go: "should adopt matching pods on creation and release no longer matching pods"
 test/e2e/apps/statefulset.go: "should perform rolling updates and roll backs of template modifications"
 test/e2e/apps/statefulset.go: "should perform canary updates and phased rolling updates of template modifications"
 test/e2e/apps/statefulset.go: "Scaling should happen in predictable order and halt if any stateful pod is unhealthy"
See instructions in test/conformance/README.md
```

File: test/e2e/apps/replica_set.go [hunk 4]
```diff
@@ -102,7 +102,14 @@ var _ = SIGDescribe("ReplicaSet", func() {
...
-	ginkgo.It("should surface a failure condition on a common issue like exceeded quota", func() {
+	/*
+	   Release : v1.16
+	   Testname: Validate ReplicaSet pod limit restrictions
+	   Description: Apply quota to the namespace, prohibiting more than two pods.
+	   When creating a ReplicaSet of 3 of a pod, ReplicaSetConditions MUST include the failure of creating more than the quota allows.
+	   Replicas of pods will be scaled down to two MUST succeed, as the quota allows this amount of replicas to exist.
+	*/
+	framework.ConformanceIt("should surface a failure condition on a common issue like exceeded quota", func() {
 		testReplicaSetConditionCheck(f)
```
Comment (johnbelamaric) [inline test/e2e/apps/replica_set.go]: The condition check https://github.com/kubernetes/kubernetes/pull/80148/files#diff-34aadb3c023dcf4d181ab896ce74d84fL225 is wrong. It should be checking that the cond.Status is true. This test as written will pass even if that condition is present but set to False.

Comment (johnbelamaric) [reply test/e2e/apps/replica_set.go]: See https://github.com/kubernetes/kubernetes/pull/74891#issuecomment-479234581 which also asks for this to check that the condition does not occur when it shouldn't (so, say, change the spec to 2 replicas and the condition should clear up).

Comment (BobyMCbobs) [reply test/e2e/apps/replica_set.go]: > The condition check https://github.com/kubernetes/kubernetes/pull/80148/files#diff-34aadb3c023dcf4d181ab896ce74d84fL225 is wrong. It should be checking that the cond.Status is true. This test as written will pass even if that condition is present but set to False.

cond.Status appears to be an [string / api.ConditionStatus](https://github.com/kubernetes/kubernetes/blob/7f1ae0e32d2d42efc243b0b3d69a95797ce67046/pkg/apis/apps/types.go#L786), not sure if the comparison for the return value can be valid.

Comment (BobyMCbobs) [reply test/e2e/apps/replica_set.go]: > The condition check https://github.com/kubernetes/kubernetes/pull/80148/files#diff-34aadb3c023dcf4d181ab896ce74d84fL225 is wrong. It should be checking that the cond.Status is true. This test as written will pass even if that condition is present but set to False.

cond.Status appears to be an [string / api.ConditionStatus](https://github.com/kubernetes/kubernetes/blob/7f1ae0e32d2d42efc243b0b3d69a95797ce67046/pkg/apis/apps/types.go#L786), not sure if the return value would be valid if trying to return as bool.

Build log (relevant lines):
```
test/e2e/apps/replica_set.go:232:22: cannot convert true (type untyped bool) to type "k8s.io/kubernetes/vendor/k8s.io/api/core/v1".ConditionStatus
test/e2e/apps/replica_set.go:232:22: invalid operation: cond.Status == true (mismatched types "k8s.io/kubernetes/vendor/k8s.io/api/core/v1".ConditionStatus and bool)
```

Comment (spiffxp) [reply test/e2e/apps/replica_set.go]: you can compare against these known values
https://github.com/kubernetes/kubernetes/blob/acaac181dc8f4c16ed0f5e47f687a6475cae49f3/staging/src/k8s.io/api/core/v1/types.go#L2304-L2308

**Comment by spiffxp** [comment]
/lgtm cancel
per @johnbelamaric's above comment, please update to check status

**Comment by devaips** [comment]
> /lgtm cancel
> per @johnbelamaric's above comment, please update to check status

ACK. 
Working on an update.

File: test/e2e/apps/replica_set.go [hunk 5]
```diff
@@ -222,7 +229,7 @@ func testReplicaSetConditionCheck(f *framework.Framework) {
 		conditions = rs.Status.Conditions
 
 		cond := replicaset.GetCondition(rs.Status, appsv1.ReplicaSetReplicaFailure)
-		return cond != nil, nil
+		return cond.Status == "True", nil
```
Comment (mgdevstack) [inline test/e2e/apps/replica_set.go]:
```suggestion
              return cond.Status == v1.ConditionTrue, nil
```

**Review by mgdevstack — CHANGES_REQUESTED** [review]

**Review by mgdevstack — CHANGES_REQUESTED** [review]
It's better to update
https://github.com/kubernetes/kubernetes/blob/0e39ca10c4ed860463e0183ed4db0b6eab4a56a9/test/e2e/apps/replica_set.go#L262
```suggestion
		return (cond == nil || cond.Status == v1.ConditionFalse), nil
```

**Comment by mgdevstack** [comment]
Above mentioned changes are also expected for `ReplicationController` based Conformance e2e residing in `e2e/apps/rc.go`
https://github.com/kubernetes/kubernetes/blob/661e4d5b3857b5268f2d6d6b79219daf2e4b5833/test/e2e/apps/rc.go#L223
https://github.com/kubernetes/kubernetes/blob/661e4d5b3857b5268f2d6d6b79219daf2e4b5833/test/e2e/apps/rc.go#L252

Looking on these expected change, it would be good to create another PR reflecting suggested changes and continuing promotion in this PR.

**Comment by devaips** [comment]
Test rewrite is happening in this PR: https://github.com/kubernetes/kubernetes/pull/81574 

Closing this PR.

**Pull request closed without merging**